# Makeover Monday Notebook
### October 7, 2024

*Part I - Data exploration*

In [1]:
import os
from os import path

import matplotlib as mpl
import polars as pl
import matplotlib.pyplot as plt

In [2]:
df = pl.read_csv(path.join(os.getcwd(), 'data.csv'))

for column in df.columns:
    print(column)

Area Names
Area Codes
Time Period
Happiness
Worthwhile
Anxiety
Life Satisfaction
Unit
Notes
Data Description


In [3]:
df.shape

(4476, 10)

In [4]:
df.head()

shape: (5, 10)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ Area      ┆ Area      ┆ Time      ┆ Happiness ┆ … ┆ Life Sati ┆ Unit      ┆ Notes     ┆ Data Des │
│ Names     ┆ Codes     ┆ Period    ┆ ---       ┆   ┆ sfaction  ┆ ---       ┆ ---       ┆ cription │
│ ---       ┆ ---       ┆ ---       ┆ f64       ┆   ┆ ---       ┆ str       ┆ str       ┆ ---      │
│ str       ┆ str       ┆ str       ┆           ┆   ┆ f64       ┆           ┆           ┆ str      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ Haringey  ┆ E09000014 ┆ April     ┆ 7.27      ┆ … ┆ 7.38      ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2014 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2015      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Mid       ┆ N09000009 ┆ April     ┆ 7.82      ┆ … ┆ 7.77      ┆ Score out ┆ 1. Data   ┆ Average  │
│ Ulster    ┆           ┆ 2013 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2014      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Swansea / ┆ W06000011 ┆ April     ┆ 7.41      ┆ … ┆ 7.53      ┆ Score out ┆ 1. Data   ┆ Average  │
│ Abertawe  ┆           ┆ 2016 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2017      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Sedgemoor ┆ E07000188 ┆ April     ┆ 7.04      ┆ … ┆ 7.64      ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2012 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2013      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Windsor   ┆ E06000040 ┆ April     ┆ 7.67      ┆ … ┆ 7.83      ┆ Score out ┆ 1. Data   ┆ Average  │
│ and Maide ┆           ┆ 2015 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│ nhead     ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2016      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [5]:
df.bottom_k(10, by='Happiness')

shape: (10, 10)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ Area      ┆ Area      ┆ Time      ┆ Happiness ┆ … ┆ Life Sati ┆ Unit      ┆ Notes     ┆ Data Des │
│ Names     ┆ Codes     ┆ Period    ┆ ---       ┆   ┆ sfaction  ┆ ---       ┆ ---       ┆ cription │
│ ---       ┆ ---       ┆ ---       ┆ f64       ┆   ┆ ---       ┆ str       ┆ str       ┆ ---      │
│ str       ┆ str       ┆ str       ┆           ┆   ┆ f64       ┆           ┆           ┆ str      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ Adur      ┆ E07000223 ┆ April     ┆ 6.16      ┆ … ┆ 7.57      ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2022 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2023      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Gedling   ┆ E07000173 ┆ April     ┆ 6.52      ┆ … ┆ 7.13      ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2020 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2021      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Harlow    ┆ E07000073 ┆ April     ┆ 6.55      ┆ … ┆ 6.79      ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2012 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2013      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Hastings  ┆ E07000062 ┆ April     ┆ 6.56      ┆ … ┆ 6.93      ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2022 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2023      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Lincoln   ┆ E07000138 ┆ April     ┆ 6.57      ┆ … ┆ 6.7       ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2022 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2023      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Eastbourn ┆ E07000061 ┆ April     ┆ 6.57      ┆ … ┆ 6.68      ┆ Score out ┆ 1. Data   ┆ Average  │
│ e         ┆           ┆ 2022 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2023      ┆           ┆   ┆           ┆           ┆ mean      ┆ personal │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ aver…     ┆ we…      │
│ Stevenage ┆ E07000243 ┆ April     ┆ 6.58      ┆ … ┆ 7.31      ┆ Score out ┆ 1. Data   ┆ Average  │
│           ┆           ┆ 2020 to   ┆           ┆   ┆           ┆ of 10     ┆ are       ┆ ratings  │
│           ┆           ┆ March     ┆           ┆   ┆           ┆           ┆ weighted  ┆ of       │
│           ┆           ┆ 2021      ┆        

Based on this, we're going to have a hard time comparing the data between categories without normalizing it. I'd also like to try regression analysis of each of the three to life happiness to try to find a correlation. Also, it looks like the columns for Unit, Notes, and Data Description are useless. Check for any unique values in those columns:

In [6]:
df_list = ['Unit', 'Notes', 'Data Description']

try_this = df.select(
    pl.col('Unit').value_counts(sort=True, name='Unit Count'),
    pl.col('Notes').value_counts(sort=True, name='Notes count'),
    pl.col('Data Description').value_counts(sort=True, name='Data Desc count'),
)

#for column in df_list:
#    unique_count_df = df.select(pl.col(column).value_counts(sort=True, name="Uniques"))
#    print(unique_count_df.head()) # We have to use print because you can't .head() in a for loop

print(type(try_this))

<class 'polars.dataframe.frame.DataFrame'>


Good to know. Is there any structure to the numbering conventions for UK zip codes? There must be...

Great question! They're called GSS codes.

[WIKIPEDIA LINK](https://en.wikipedia.org/wiki/GSS_coding_system)

Jackpot. Those first three characters speak to the "level of geography". As I understand it, this represents the type of municipality represented by the "Area Names" value. The way the data was gathered filtered out all non-municipal codes already.

I think we need a new column.

In [7]:
zones = []

for entry in df.rows():
    zones.append(entry[1][0:3])

zones = pl.Series("Zones", zones)
df.insert_column(2, zones)
df.head()

shape: (5, 11)
┌────────────┬────────────┬───────┬────────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ Area Names ┆ Area Codes ┆ Zones ┆ Time       ┆ … ┆ Life Sati ┆ Unit      ┆ Notes     ┆ Data Desc │
│ ---        ┆ ---        ┆ ---   ┆ Period     ┆   ┆ sfaction  ┆ ---       ┆ ---       ┆ ription   │
│ str        ┆ str        ┆ str   ┆ ---        ┆   ┆ ---       ┆ str       ┆ str       ┆ ---       │
│            ┆            ┆       ┆ str        ┆   ┆ f64       ┆           ┆           ┆ str       │
╞════════════╪════════════╪═══════╪════════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ Haringey   ┆ E09000014  ┆ E09   ┆ April 2014 ┆ … ┆ 7.38      ┆ Score out ┆ 1. Data   ┆ Average   │
│            ┆            ┆       ┆ to March   ┆   ┆           ┆ of 10     ┆ are       ┆ ratings   │
│            ┆            ┆       ┆ 2015       ┆   ┆           ┆           ┆ weighted  ┆ of        │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ mean      ┆ personal  │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ aver…     ┆ we…       │
│ Mid Ulster ┆ N09000009  ┆ N09   ┆ April 2013 ┆ … ┆ 7.77      ┆ Score out ┆ 1. Data   ┆ Average   │
│            ┆            ┆       ┆ to March   ┆   ┆           ┆ of 10     ┆ are       ┆ ratings   │
│            ┆            ┆       ┆ 2014       ┆   ┆           ┆           ┆ weighted  ┆ of        │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ mean      ┆ personal  │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ aver…     ┆ we…       │
│ Swansea /  ┆ W06000011  ┆ W06   ┆ April 2016 ┆ … ┆ 7.53      ┆ Score out ┆ 1. Data   ┆ Average   │
│ Abertawe   ┆            ┆       ┆ to March   ┆   ┆           ┆ of 10     ┆ are       ┆ ratings   │
│            ┆            ┆       ┆ 2017       ┆   ┆           ┆           ┆ weighted  ┆ of        │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ mean      ┆ personal  │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ aver…     ┆ we…       │
│ Sedgemoor  ┆ E07000188  ┆ E07   ┆ April 2012 ┆ … ┆ 7.64      ┆ Score out ┆ 1. Data   ┆ Average   │
│            ┆            ┆       ┆ to March   ┆   ┆           ┆ of 10     ┆ are       ┆ ratings   │
│            ┆            ┆       ┆ 2013       ┆   ┆           ┆           ┆ weighted  ┆ of        │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ mean      ┆ personal  │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ aver…     ┆ we…       │
│ Windsor    ┆ E06000040  ┆ E06   ┆ April 2015 ┆ … ┆ 7.83      ┆ Score out ┆ 1. Data   ┆ Average   │
│ and        ┆            ┆       ┆ to March   ┆   ┆           ┆ of 10     ┆ are       ┆ ratings   │
│ Maidenhead ┆            ┆       ┆ 2016       ┆   ┆           ┆           ┆ weighted  ┆ of        │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ mean      ┆ personal  │
│            ┆            ┆       ┆            ┆   ┆           ┆           ┆ aver…     ┆ we…       │
└────────────┴────────────┴───────┴────────────┴───┴───────────┴───────────┴───────────┴───────────┘

I want to see how many unique zones we have, and decide which types they are

In [8]:
zone_count = df.select(pl.col('Zones').value_counts(sort=True, name="Codes"))
print(zone_count)

# I don't know how to get this to display as anything other than the counted data contained in a set of brackets

shape: (7, 1)
┌──────────────┐
│ Zones        │
│ ---          │
│ struct[2]    │
╞══════════════╡
│ {"E07",2172} │
│ {"E06",696}  │
│ {"E08",432}  │
│ {"E09",396}  │
│ {"S12",384}  │
│ {"W06",264}  │
│ {"N09",132}  │
└──────────────┘


This corresponds to :
* E07 -- 2172 -- Non-Metropolitan District
* E06 -- 696 -- Unitary Authority
* E08 -- 432 -- Metropolitan Bureau
* E09 -- 396 -- London Burough
* S12 -- 384 -- Council Areas
* W06 -- 264 -- Unitary Authority
* N09 -- 132 -- Local Government Districts

Our next big challenge is that the timeframes these data were taken on are somewhat different. Let's see...

In [19]:
zone_count = df.select(pl.col('Area Names').value_counts(sort=True, name="Stuff"))

print(zone_count)

# The below will show the whole thing so you can double check, but each town has 12 rows.

#with pl.Config(tbl_rows=-1):
#    print(zone_count)

shape: (308, 1)
┌───────────────────────────────┐
│ Area Names                    │
│ ---                           │
│ struct[2]                     │
╞═══════════════════════════════╡
│ {"Haringey",12}               │
│ {"Sedgemoor",12}              │
│ {"Windsor and Maidenhead",12} │
│ {"Nuneaton and Bedworth",12}  │
│ {"Hyndburn",12}               │
│ …                             │
│ {"South Norfolk",12}          │
│ {"Plymouth",12}               │
│ {"Brent",12}                  │
│ {"City of London",12}         │
│ {"North Tyneside",12}         │
└───────────────────────────────┘


*Part II - Visualization*

We'll visualize the data before doing regression analysis to see whether it's even worth it. 

We're going to look into whether these categorized area codes and reported happiness are independent. I'd also like to look at how ratings for happiness have changed over time, and whether certain government types may lead to a more or less happy populace.

Unfortunately, since I'm bringing in government structures I'm going to drop the Scottish, Welsh, and Irish zones from the analysis. They may share, as an example, the Unitary Authority feature, but the national level laws governing those authorities will be different.

Specifically, I'd like to look at:
* Unitary Authority vs. non- Unitary Authority
* Metropolitan vs. non- Metropolitan
* London Metropolitan vs. non- London Metropolitan
* Of E07-E09, which has seen the most change in reported happiness rating?

Our first step is to restructure the dataframes. Let's sort first by area codes, then area names, then time period. Then... we can cheat and split the dataframe into groups of 12 by row index.

In [16]:
# I don't know whether it's good practice to constantly overwrite the dataframe, but I don't want it living in memory too long. It can be painful to live with that sorrow.

df = df.filter(
    ~pl.col('Zones').is_in(['W06', 'S12', 'N09']),
)

sorted_df = df.sort(['Zones', 'Area Names', 'Time Period'])

# The below code will print a massive version of the dataframe. I used it to check the sorting consistency.
# It's large enough to be inconvenient, thus it's commented out, but left available.

#with pl.Config(tbl_rows=100, tbl_cols=-1):
#    print(sorted_df)

The number of rows removed should be equal to the sum of:
* S12 -- 384 -- Council Areas
* W06 -- 264 -- Unitary Authority
* N09 -- 132 -- Local Government Districts

In [14]:
sorted_df.height == 4476 - (384 + 264 + 132) # I know you can distribute the negative Mr. Smarty Pants, but math is a visual game

True

Nowadays I'm also tired of seeing those extra columns

In [20]:
df = df.select(
    pl.col(['Area Names', 'Area Codes', 'Zones', 'Time Period', 'Happiness', 'Worthwhile', 'Anxiety', 'Life Satisfaction'])
)

df.head()

shape: (5, 8)
┌──────────────┬────────────┬───────┬─────────────┬───────────┬────────────┬─────────┬─────────────┐
│ Area Names   ┆ Area Codes ┆ Zones ┆ Time Period ┆ Happiness ┆ Worthwhile ┆ Anxiety ┆ Life Satisf │
│ ---          ┆ ---        ┆ ---   ┆ ---         ┆ ---       ┆ ---        ┆ ---     ┆ action      │
│ str          ┆ str        ┆ str   ┆ str         ┆ f64       ┆ f64        ┆ f64     ┆ ---         │
│              ┆            ┆       ┆             ┆           ┆            ┆         ┆ f64         │
╞══════════════╪════════════╪═══════╪═════════════╪═══════════╪════════════╪═════════╪═════════════╡
│ Haringey     ┆ E09000014  ┆ E09   ┆ April 2014  ┆ 7.27      ┆ 7.63       ┆ 3.0     ┆ 7.38        │
│              ┆            ┆       ┆ to March    ┆           ┆            ┆         ┆             │
│              ┆            ┆       ┆ 2015        ┆           ┆            ┆         ┆             │
│ Sedgemoor    ┆ E07000188  ┆ E07   ┆ April 2012  ┆ 7.04      ┆ 7.76       ┆ 3.46    ┆ 7.64        │
│              ┆            ┆       ┆ to March    ┆           ┆            ┆         ┆             │
│              ┆            ┆       ┆ 2013        ┆           ┆            ┆         ┆             │
│ Windsor and  ┆ E06000040  ┆ E06   ┆ April 2015  ┆ 7.67      ┆ 7.97       ┆ 2.98    ┆ 7.83        │
│ Maidenhead   ┆            ┆       ┆ to March    ┆           ┆            ┆         ┆             │
│              ┆            ┆       ┆ 2016        ┆           ┆            ┆         ┆             │
│ Nuneaton and ┆ E07000219  ┆ E07   ┆ April 2012  ┆ 7.26      ┆ 7.64       ┆ 3.09    ┆ 7.7         │
│ Bedworth     ┆            ┆       ┆ to March    ┆           ┆            ┆         ┆             │
│              ┆            ┆       ┆ 2013        ┆           ┆            ┆         ┆             │
│ Hyndburn     ┆ E07000120  ┆ E07   ┆ April 2018  ┆ 7.89      ┆ 7.73       ┆ 2.7     ┆ 7.99        │
│              ┆            ┆       ┆ to March    ┆           ┆            ┆         ┆             │
│              ┆            ┆       ┆ 2019        ┆           ┆            ┆         ┆             │
└──────────────┴────────────┴───────┴─────────────┴───────────┴────────────┴─────────┴─────────────┘